File: app/src/main/java/com/example/starwars/adapter/FilmsAdapter.java
package com.example.starwars.adapter;

import android.content.Context;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.starwars.R;
import com.example.starwars.StarWarsApp;
import com.example.starwars.model.Film;

import java.util.ArrayList;

public class FilmsAdapter extends RecyclerView.Adapter<FilmsAdapter.ViewHolder> {

    private static final String IMAGE_BASE_URL = "https://image.tmdb.org/t/p/w342/";

    private ArrayList<Film> mFilms;
    private LayoutInflater inflater;
    private FilmClickListener filmClickListener;

    public FilmsAdapter(ArrayList<Film> films, Context context, FilmClickListener filmClickListener) {
        this.mFilms = films;
        this.inflater = LayoutInflater.from(context);
        this.filmClickListener = filmClickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.film_list_item, parent, false);
        return new ViewHolder(view);
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Film film = mFilms.get(position);
        holder.FilmTitleTextView.setText(film.getTitle());
        holder.FilmInfoTextView.setText("Average vote: " + film.getVoteAverage());
        Glide.with(StarWarsApp.getContext())
                .load(IMAGE_BASE_URL + film.getPosterPath())
                .error(R.drawable.ic_launcher_background)
                .into(holder.FilmIconImageView);
    }

    @Override
    public int getItemCount() {
        return mFilms.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView FilmTitleTextView;
        TextView FilmInfoTextView;
        ImageView FilmIconImageView;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
            FilmTitleTextView = itemView.findViewById(R.id.film_item_title);
            FilmInfoTextView = itemView.findViewById(R.id.film_item_info);
            FilmIconImageView = itemView.findViewById(R.id.film_item_icon);
        }

        @Override
        public void onClick(View v) {
            filmClickListener.onItemClick(getAdapterPosition());
        }
    }

    public interface FilmClickListener {
        void onItemClick(int position);
    }

    public void setFilms(ArrayList<Film> filmList) {
        mFilms = filmList;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/starwars/adapter/FilmsAdapter.kt
package com.example.starwars.adapter

import android.content.Context
import com.example.starwars.model.Film
import com.example.starwars.adapter.FilmsAdapter.FilmClickListener
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.example.starwars.R
import androidx.annotation.RequiresApi
import android.os.Build
import android.util.Log
import android.view.View
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.example.starwars.StarWarsApp
import com.example.starwars.adapter.FilmsAdapter
import android.widget.TextView
import com.example.starwars.activity.FilmListActivity.Companion.TAG
import com.example.starwars.databinding.FilmListItemBinding
import java.util.ArrayList

class FilmsAdapter(
    private var mFilms: ArrayList<Film>,
    filmClickListener: FilmClickListener
) : RecyclerView.Adapter<FilmsAdapter.ViewHolder>() {
    private val filmClickListener: FilmClickListener = filmClickListener

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = FilmListItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        with(holder) {
            val film = mFilms[position]
            binding.filmItemTitle.text = film.title
            // TODO move "Average vote: " to string resources
            binding.filmItemInfo.text = "Average vote: " + film.voteAverage

            Glide.with(holder.itemView.context)
                .load(IMAGE_BASE_URL + film.posterPath)
                .error(R.drawable.ic_launcher_background)
                .into(binding.filmItemIcon)

            holder.itemView.setOnClickListener { filmClickListener.onItemClick(bindingAdapterPosition) }
        }
    }

    override fun getItemCount(): Int {
        return mFilms.size
    }

    inner class ViewHolder(val binding: FilmListItemBinding)
        : RecyclerView.ViewHolder(binding.root)

    interface FilmClickListener {
        fun onItemClick(position: Int)
    }

    fun setFilms(filmList: ArrayList<Film>) {
        mFilms = filmList
        Log.d(TAG, "setFilms: inside adapter")
        notifyDataSetChanged()
    }

    companion object {
        private const val IMAGE_BASE_URL = "https://image.tmdb.org/t/p/w342/"
    }
}